File: order-count-endpoint-mismatch/node/reconcile-order-counts.js
/**
 * Reconcile a BigCommerce order count mismatch between /v2/orders/count and /v2/orders.
 *
 * GET /v2/orders/count and GET /v2/orders both accept status_id, min_date_created,
 * max_date_created, and customer_id, and both apply an implicit default scope when
 * status_id is omitted. Incomplete orders (status_id 0, abandoned at payment) are
 * commonly excluded from an unfiltered count's default scope but still appear in an
 * unfiltered pagination scan, so a script calling one endpoint with no filters and
 * the other with a different filter set ends up comparing two different result
 * sets. A secondary cause is timing: count is a point-in-time snapshot, while a
 * multi-page scan can take seconds to minutes on a large store. This job sums
 * per-status counts across all 15 status_id values, fully paginates the order list
 * with the same filters, reconciles the two totals bucket by bucket, and re-checks
 * the count snapshot after pagination to rule out concurrency drift. It only ever
 * reports. It never deletes or modifies an order based on a count mismatch alone.
 *
 * Guide: https://www.allanninal.dev/bigcommerce/order-count-endpoint-mismatch/
 */
import { pathToFileURL } from "node:url";

const STORE_HASH = process.env.BIGCOMMERCE_STORE_HASH || "example_hash";
const ACCESS_TOKEN = process.env.BIGCOMMERCE_ACCESS_TOKEN || "bc_dummy";
const API_BASE = `https://api.bigcommerce.com/stores/${STORE_HASH}/v2`;
const MIN_DATE_CREATED = process.env.MIN_DATE_CREATED || null;
const DRY_RUN = (process.env.DRY_RUN || "true").toLowerCase() === "true";

const ALL_STATUS_IDS = Array.from({ length: 15 }, (_, i) => i); // 0 Incomplete .. 14 Partially Refunded
const PAGE_LIMIT = 250;

const HEADERS = {
  "X-Auth-Token": ACCESS_TOKEN,
  "Content-Type": "application/json",
  Accept: "application/json",
};

/**
 * Pure comparison. No network, no side effects.
 *
 * countEndpointTotals maps statusId -> count returned by
 * GET /v2/orders/count?status_id={id} for each of the 15 status_id values.
 * paginatedOrderStatusIds is the flat list of status_id values collected by
 * fully paginating GET /v2/orders with the same filters. Returns a plain report
 * with the two grand totals, a per-status delta map, the list of status_id
 * values where the two disagree, and whether every bucket balances.
 */
export function reconcileOrderCounts(countEndpointTotals, paginatedOrderStatusIds) {
  const paginatedCounts = new Map();
  for (const statusId of paginatedOrderStatusIds) {
    paginatedCounts.set(statusId, (paginatedCounts.get(statusId) || 0) + 1);
  }

  const allStatusIds = new Set([
    ...Object.keys(countEndpointTotals).map(Number),
    ...paginatedCounts.keys(),
  ]);

  const perStatusDeltas = {};
  for (const statusId of allStatusIds) {
    const expected = countEndpointTotals[statusId] || 0;
    const actual = paginatedCounts.get(statusId) || 0;
    perStatusDeltas[statusId] = expected - actual;
  }

  const mismatchedStatusIds = Object.keys(perStatusDeltas)
    .map(Number)
    .filter((sid) => perStatusDeltas[sid] !== 0)
    .sort((a, b) => a - b);

  return {
    totalCountEndpoint: Object.values(countEndpointTotals).reduce((a, b) => a + b, 0),
    totalPaginated: paginatedOrderStatusIds.length,
    perStatusDeltas,
    mismatchedStatusIds,
    isConsistent: mismatchedStatusIds.length === 0,
  };
}

async function bcGet(path, params = {}) {
  const url = new URL(`${API_BASE}${path}`);
  for (const [key, value] of Object.entries(params)) {
    if (value !== undefined && value !== null) url.searchParams.set(key, value);
  }
  const res = await fetch(url, { headers: HEADERS });
  if (!res.ok) throw new Error(`BigCommerce ${res.status}`);
  const text = await res.text();
  return text ? JSON.parse(text) : [];
}

async function countByStatus() {
  const totals = {};
  for (const statusId of ALL_STATUS_IDS) {
    const params = { status_id: statusId };
    if (MIN_DATE_CREATED) params.min_date_created = MIN_DATE_CREATED;
    const body = await bcGet("/orders/count", params);
    totals[statusId] = body.count || 0;
  }
  return totals;
}

async function paginateAllOrderStatusIds() {
  const statusIds = [];
  let page = 1;
  while (true) {
    const params = { page, limit: PAGE_LIMIT, sort: "id:asc" };
    if (MIN_DATE_CREATED) params.min_date_created = MIN_DATE_CREATED;
    const orders = await bcGet("/orders", params);
    if (!orders.length) return statusIds;
    for (const order of orders) statusIds.push(order.status_id);
    if (orders.length < PAGE_LIMIT) return statusIds;
    page += 1;
  }
}

function logReport(label, report) {
  console.log(`[${label}] unfiltered_count=${report.totalCountEndpoint} paginated_total=${report.totalPaginated} is_consistent=${report.isConsistent}`);
  if (!report.isConsistent) {
    for (const statusId of report.mismatchedStatusIds) {
      console.warn(`[${label}] status_id=${statusId} delta=${report.perStatusDeltas[statusId]}`);
    }
  }
}

export async function run() {
  console.log(`Fetching per-status counts before pagination (DRY_RUN=${DRY_RUN}, report only, no writes).`);
  const countsBefore = await countByStatus();

  const paginatedStatusIds = await paginateAllOrderStatusIds();

  console.log("Fetching per-status counts again after pagination to check for concurrency drift.");
  const countsAfter = await countByStatus();

  const reportBefore = reconcileOrderCounts(countsBefore, paginatedStatusIds);
  const reportAfter = reconcileOrderCounts(countsAfter, paginatedStatusIds);

  logReport("pre-scan snapshot", reportBefore);
  logReport("post-scan snapshot", reportAfter);

  if (reportBefore.isConsistent && reportAfter.isConsistent) {
    console.log("Consistent. Counts and pagination agree across all status_id buckets.");
    return;
  }

  if (!reportBefore.isConsistent && reportAfter.isConsistent) {
    console.log("Mismatch resolved by the post-scan snapshot. Likely concurrency drift during the scan window.");
    return;
  }

  console.warn(
    `Persistent mismatch after re-checking the count snapshot. mismatchedStatusIds=${JSON.stringify(reportAfter.mismatchedStatusIds)}. ` +
    `This is a report for a human, escalate to BigCommerce support with store_hash=${STORE_HASH}, ` +
    `min_date_created=${MIN_DATE_CREATED}, and the mismatched status_id list. No orders were modified.`
  );
}

if (import.meta.url === pathToFileURL(process.argv[1]).href) {
  run().catch((err) => { console.error(err); process.exit(1); });
}

File: order-count-endpoint-mismatch/python/reconcile_order_counts.py
"""Reconcile a BigCommerce order count mismatch between /v2/orders/count and /v2/orders.

GET /v2/orders/count and GET /v2/orders both accept status_id, min_date_created,
max_date_created, and customer_id, and both apply an implicit default scope when
status_id is omitted. Incomplete orders (status_id 0, abandoned at payment) are
commonly excluded from an unfiltered count's default scope but still appear in an
unfiltered pagination scan, so a script calling one endpoint with no filters and
the other with a different filter set ends up comparing two different result
sets. A secondary cause is timing: count is a point-in-time snapshot, while a
multi-page scan can take seconds to minutes on a large store. This job sums
per-status counts across all 15 status_id values, fully paginates the order list
with the same filters, reconciles the two totals bucket by bucket, and re-checks
the count snapshot after pagination to rule out concurrency drift. It only ever
reports. It never deletes or modifies an order based on a count mismatch alone.

Guide: https://www.allanninal.dev/bigcommerce/order-count-endpoint-mismatch/
"""
import os
import logging
from collections import Counter
from dataclasses import dataclass

import requests

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("reconcile_order_counts")

STORE_HASH = os.environ.get("BIGCOMMERCE_STORE_HASH", "example_hash")
ACCESS_TOKEN = os.environ.get("BIGCOMMERCE_ACCESS_TOKEN", "bc_dummy")
API_BASE = f"https://api.bigcommerce.com/stores/{STORE_HASH}/v2"
MIN_DATE_CREATED = os.environ.get("MIN_DATE_CREATED") or None
DRY_RUN = os.environ.get("DRY_RUN", "true").lower() == "true"

ALL_STATUS_IDS = list(range(15))  # 0 Incomplete .. 14 Partially Refunded
PAGE_LIMIT = 250

HEADERS = {
    "X-Auth-Token": ACCESS_TOKEN,
    "Content-Type": "application/json",
    "Accept": "application/json",
}


def bc_get(path, params=None):
    r = requests.get(f"{API_BASE}{path}", headers=HEADERS, params=params or {}, timeout=30)
    r.raise_for_status()
    if not r.text:
        return []
    return r.json()


@dataclass
class ReconciliationReport:
    total_count_endpoint: int
    total_paginated: int
    per_status_deltas: dict
    mismatched_status_ids: list
    is_consistent: bool


def reconcile_order_counts(count_endpoint_totals: dict, paginated_order_status_ids: list) -> ReconciliationReport:
    """Pure comparison. No network, no side effects.

    count_endpoint_totals maps status_id -> count returned by
    GET /v2/orders/count?status_id={id} for each of the 15 status_id values.
    paginated_order_status_ids is the flat list of status_id values collected by
    fully paginating GET /v2/orders with the same filters. Returns a plain report
    with the two grand totals, a per-status delta map, the list of status_id
    values where the two disagree, and whether every bucket balances.
    """
    paginated_counts = Counter(paginated_order_status_ids)
    all_status_ids = set(count_endpoint_totals) | set(paginated_counts)

    per_status_deltas = {}
    for status_id in all_status_ids:
        expected = count_endpoint_totals.get(status_id, 0)
        actual = paginated_counts.get(status_id, 0)
        per_status_deltas[status_id] = expected - actual

    mismatched_status_ids = sorted(sid for sid, delta in per_status_deltas.items() if delta != 0)

    return ReconciliationReport(
        total_count_endpoint=sum(count_endpoint_totals.values()),
        total_paginated=len(paginated_order_status_ids),
        per_status_deltas=per_status_deltas,
        mismatched_status_ids=mismatched_status_ids,
        is_consistent=all(delta == 0 for delta in per_status_deltas.values()),
    )


def count_by_status():
    totals = {}
    for status_id in ALL_STATUS_IDS:
        params = {"status_id": status_id}
        if MIN_DATE_CREATED:
            params["min_date_created"] = MIN_DATE_CREATED
        body = bc_get("/orders/count", params)
        totals[status_id] = body.get("count", 0)
    return totals


def paginate_all_order_status_ids():
    status_ids = []
    page = 1
    while True:
        params = {"page": page, "limit": PAGE_LIMIT, "sort": "id:asc"}
        if MIN_DATE_CREATED:
            params["min_date_created"] = MIN_DATE_CREATED
        orders = bc_get("/orders", params)
        if not orders:
            return status_ids
        for order in orders:
            status_ids.append(order["status_id"])
        if len(orders) < PAGE_LIMIT:
            return status_ids
        page += 1


def log_report(label, report):
    log.info("[%s] unfiltered_count=%s paginated_total=%s is_consistent=%s",
              label, report.total_count_endpoint, report.total_paginated, report.is_consistent)
    if not report.is_consistent:
        for status_id in report.mismatched_status_ids:
            log.warning("[%s] status_id=%s delta=%s", label, status_id, report.per_status_deltas[status_id])


def run():
    log.info("Fetching per-status counts before pagination (DRY_RUN=%s, report only, no writes).", DRY_RUN)
    counts_before = count_by_status()

    paginated_status_ids = paginate_all_order_status_ids()

    log.info("Fetching per-status counts again after pagination to check for concurrency drift.")
    counts_after = count_by_status()

    report_before = reconcile_order_counts(counts_before, paginated_status_ids)
    report_after = reconcile_order_counts(counts_after, paginated_status_ids)

    log_report("pre-scan snapshot", report_before)
    log_report("post-scan snapshot", report_after)

    if report_before.is_consistent and report_after.is_consistent:
        log.info("Consistent. Counts and pagination agree across all status_id buckets.")
        return

    if not report_before.is_consistent and report_after.is_consistent:
        log.info("Mismatch resolved by the post-scan snapshot. Likely concurrency drift during the scan window.")
        return

    log.warning(
        "Persistent mismatch after re-checking the count snapshot. mismatched_status_ids=%s. "
        "This is a report for a human, escalate to BigCommerce support with store_hash=%s, "
        "min_date_created=%s, and the mismatched status_id list. No orders were modified.",
        report_after.mismatched_status_ids, STORE_HASH, MIN_DATE_CREATED,
    )


if __name__ == "__main__":
    run()
